django/core/management/commands/migrate.py

```diff
@@ -259,63 +259,54 @@ def migration_progress_callback(self, action, migration=None, fake=False):
+            converter = connection.introspection.table_name_converter
+            # Note that if a model is unmanaged we short-circuit and never try to install it
+            return not (
+                (converter(opts.db_table) in tables) or
+                (opts.auto_created and converter(opts.auto_created._meta.db_table) in tables)
             )
 
-            # Create the tables for each model
-            if self.verbosity >= 1:
-                self.stdout.write("  Creating tables...\n")
-            with transaction.atomic(using=connection.alias, savepoint=connection.features.can_rollback_ddl):
-                deferred_sql = []
-                for app_name, model_list in manifest.items():
-                    for model in model_list:
-                        if not model._meta.can_migrate(connection):
-                            continue
-                        if self.verbosity >= 3:
-                            self.stdout.write(
-                                "    Processing %s.%s model\n" % (app_name, model._meta.object_name)
-                            )
-                        with connection.schema_editor() as editor:
-                            if self.verbosity >= 1:
-                                self.stdout.write("    Creating table %s\n" % model._meta.db_table)
-                            editor.create_model(model)
-                            deferred_sql.extend(editor.deferred_sql)
-                            editor.deferred_sql = []
-                        created_models.add(model)
+        manifest = OrderedDict(
+            (app_name, list(filter(model_installed, model_list)))
+            for app_name, model_list in all_models
+        )
 
-                if self.verbosity >= 1:
-                    self.stdout.write("    Running deferred SQL...\n")
-                with connection.schema_editor() as editor:
-                    for statement in deferred_sql:
-                        editor.execute(statement)
-        finally:
-            cursor.close()
+        # Create the tables for each model
+        if self.verbosity >= 1:
+            self.stdout.write("  Creating tables...\n")
+        with connection.schema_editor() as editor:
+            for app_name, model_list in manifest.items():
+                for model in model_list:
+                    if not model._meta.can_migrate(connection):
+                        continue
+                    if self.verbosity >= 3:
+                        self.stdout.write(
+                            "    Processing %s.%s model\n" % (app_name, model._meta.object_name)
+                        )
+                    if self.verbosity >= 1:
+                        self.stdout.write("    Creating table %s\n" % model._meta.db_table)
+                    editor.create_model(model)
+                    created_models.add(model)
+
+            if self.verbosity >= 1:
+                self.stdout.write("    Running deferred SQL...\n")
```

Comment:
This cleans up two things.

1. Keep the `connection.cursor()` opened only for the duration of table introspection.
2. Use `connection.schema_editor()` instead of `atomic()` which will [make sure to use `atomic()` only if `connection.features.can_rollback_ddl`](https://github.com/django/django/blob/96181080bae66ee4f4e6d52bddcd57fa420a929b/django/db/backends/base/schema.py#L72-L86).